node/silkworm/stagedsync/stage_execution.cpp

```diff
@@ -108,52 +109,27 @@ StageResult Execution::forward(db::RWTxn& txn) {
     return is_stopping() ? StageResult::kAborted : StageResult::kSuccess;
 }
 
-std::queue<Block> Execution::prefetch_blocks(db::RWTxn& txn, BlockNum from, BlockNum to, size_t max_blocks) {
+void Execution::prefetch_blocks(db::RWTxn& txn, const BlockNum from, const BlockNum to, const size_t max_blocks) {
+    SILKWORM_ASSERT(from <= to);
+
     std::unique_ptr<StopWatch> sw;
     if (log::test_verbosity(log::Level::kTrace)) {
         sw = std::make_unique<StopWatch>(/*auto_start=*/true);
     }
 
-    std::queue<Block> ret{};
-    auto hashes_table{db::open_cursor(*txn, db::table::kCanonicalHashes)};
-    auto key{db::block_key(from)};
-    auto data{hashes_table.find(db::to_slice(key), true)};
-    while (data.done) {
-        BlockNum reached_block_num{endian::load_big_u64(static_cast<const uint8_t*>(data.key.data()))};
-        if (reached_block_num != from) {
-            throw std::runtime_error("Bad canonical header sequence: expected " + std::to_string(from) + " got " +
-                                     std::to_string(reached_block_num));
-        }
-
-        Bytes block_key(8 + kHashLength, '\0');
-        std::memcpy(&block_key[0], data.key.data(), 8);
-        std::memcpy(&block_key[8], data.value.data(), kHashLength);
-
-        Block block{};
-        auto raw_header{db::read_header_raw(*txn, block_key)};
-        if (raw_header.empty()) {
-            throw std::runtime_error("Unable to load block header for block " + std::to_string(from));
-        }
-        ByteView encoded_header{raw_header.data(), raw_header.length()};
-        rlp::success_or_throw(rlp::decode(encoded_header, block.header));
-
-        if (!db::read_body(*txn, block_key, /*read_senders=*/true, block)) {
-            throw std::runtime_error("Unable to load block body for block " + std::to_string(from));
-        }
-        ret.push(block);
+    const size_t n{std::min(static_cast<size_t>(to - from + 1), max_blocks)};
+    prefetched_blocks_.resize(n);
 
-        if (from == to || ret.size() >= max_blocks) {
-            break;
+    for (size_t i{0}; i < n; ++i) {
+        if (!db::read_block(*txn, from + i, /*read_senders=*/true, prefetched_blocks_[i])) {
+            throw std::runtime_error("Bad canonical header sequence: missing block " + std::to_string(from + i));
```

Comment:
This imho worsen things. Original implementation uses one cursor to read canonical hashes sequentially (only one lookup on begin of loop and then next key, next key ...) and invokes `db::read_body` by key of block_num + hash which eventually opens 1 cursor and does 1 lookup on : BlockBodies, BlockTransactions
Here instead we force 1 lookup for each block on CanonicalHeaders too

Reply:
Fair enough. I've re-introduced the batch CanonicalHeaders logic in `read_blocks` in `access_layer`.